### PR TITLE
[D] Update spectrogram to use complex

### PR DESCRIPTION
**Pull request by mthrok**
This PR is for getting signal on complex tensor adaptation, similar to #864

The internal changes should go into #1013, as it resolves the issue with backward pass of the Spectrogram. Then this PR will be about adding `return_complex` arguments for the migration.

### CODE REVIEW

File: torchaudio/functional.py [hunk 1]
```diff
@@ -75,6 +76,8 @@ def spectrogram(
         ``n_fft // 2 + 1`` and ``n_fft`` is the number of
         Fourier bins, and time is the number of window hops (n_frame).
     """
+    if power is not None and return_complex:
+        raise ValueError('When `power` is provided, `return_complex` must be False.')
```
Comment (anjali411) [inline torchaudio/functional.py:80]: why is that?

File: torchaudio/functional.py [hunk 2]
```diff
@@ -86,16 +89,21 @@ def spectrogram(
...
     # unpack batch
-    spec_f = spec_f.reshape(shape[:-1] + spec_f.shape[-3:])
+    spec_f = spec_f.reshape(shape[:-1] + spec_f.shape[-2:])
 
     if normalized:
         spec_f /= window.pow(2.).sum().sqrt()
     if power is not None:
-        spec_f = complex_norm(spec_f, power=power)
+        spec_f = spec_f.abs().pow(power)
+    elif not return_complex:
```
Comment (anjali411) [inline torchaudio/functional.py:105]: shouldn't it be a separate if statement?

File: torchaudio/functional.py (hunk 1 above)
Comment (mthrok) [reply torchaudio/functional.py:80]: In the original implementation, when the `power` argument is provided, the returned signal changes from pseudo-complex Tensor to plain real Tensor. (as the spectrogram is passed to `complex_norm` function which is equivalent of `abs` of complex type tensor).

We could still return complex type with all the imaginary parts equal zero, but I thought there is not much point of that. I am open to changing it though.

File: torchaudio/functional.py (hunk 2 above)
Comment (mthrok) [reply torchaudio/functional.py:105]: As I mentioned in the previous reply, when `power` is provided, this function returns the power of spectrogram, which is real value. So `return_complex` only applies when `power` is not applied (the function is returning complex-valued tensor).

**Comment by mthrok** [comment]
@anjali411 

Do you have a suggestion for the tests to be added with respect to the complex dtype?

File: torchaudio/functional.py (hunk 1 above)
Comment (anjali411) [reply torchaudio/functional.py:80]: `torch.abs` returns a floating point tensor though so I don't see any reason why we should change it

Comment (anjali411) [reply torchaudio/functional.py:80]: However for performance reasons we should add a special case for power 1 (not call pow), and pow 2 (check if torch.view_as_real(x).pow(2).sum(-1) is performance wise better that `x.abs().pow(2)`).

Comment (mthrok) [reply torchaudio/functional.py:80]: Okay, thanks.

> However for performance reasons we should add a special case for power 1 (not call pow), and pow 2 (check if torch.view_as_real(x).pow(2).sum(-1) is performance wise better that `x.abs().pow(2)`).

I will look into this.

Comment (mthrok) [reply torchaudio/functional.py:80]: I got the following result, `torch.view_as_real(x).pow(2).sum(-1)` is slightly faster on CUDA.

| Method                                            | CPU                                   | GPU                              |
|---------------------------------|------------------------------------------|------------------------------------------|
| `x.abs().pow(2)`                                          | 100 loops, best of 5: 3.17 msec per loop | 100 loops, best of 5: 48.4 usec per loop |
| `torch.view_as_real(x).pow(2).sum(-1)` | 100 loops, best of 5: 28.8 msec per loop | 100 loops, best of 5: 41.7 usec per loop |

<details> <summary>script</summary>

```
OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat)
""" """
x.abs().pow(2);
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat)
""" """
torch.view_as_real(x).pow(2).sum(-1)
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat, device='cuda')
""" """
x.abs().pow(2);
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -n 100 -r 5 -s """
import torch;
import torch.fft;
x = torch.zeros(1, 32*44100, dtype=torch.cfloat, device='cuda')
""" """
torch.view_as_real(x).pow(2).sum(-1)
"""
```

</details>

Comment (mthrok) [reply torchaudio/functional.py:80]: @anjali411 Updated the code based on the result.

File: torchaudio/functional.py [hunk 3]
```diff
@@ -75,6 +76,8 @@ def spectrogram(
         ``n_fft // 2 + 1`` and ``n_fft`` is the number of
         Fourier bins, and time is the number of window hops (n_frame).
     """
+    if power is not None and return_complex:
```
Comment (anjali411) [inline torchaudio/functional.py:79]: based on the discussion below, we should get rid of this

Comment (mthrok) [reply torchaudio/functional.py:79]: Okay. So the suggestion is that if power is provided, we return real-type tensor regardless of the `return_complex` value?

File: torchaudio/functional/functional.py [hunk 4]
```diff
@@ -91,17 +94,30 @@ def spectrogram(
...
     if power is not None:
-        spec_f = complex_norm(spec_f, power=power)
-
+        if power == 1.0:
+            return spec_f.abs()
+        if power == 2.0 and spec_f.is_cuda:
+            return torch.view_as_real(spec_f).pow(power).sum(-1)
+        return spec_f.abs().pow(power)
+    if not return_complex:
+        return torch.view_as_real(spec_f)
```
Comment (vincentqb) [inline torchaudio/functional/functional.py:120]: [same](https://github.com/pytorch/audio/pull/1013/files#r520215197)

**Pull request closed without merging**